compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::span::Span;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind},
+    cooked::TokenKind,
+    Delim,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    pub span: Span,
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
```

Comment:
If the original document uses tabs or a different indent width should we respect that? I think VS Code provides these options via `FormattingOptions` https://code.visualstudio.com/api/references/vscode-api#FormattingOptions

Reply:
I had thought of implementing that as future work for the formatter, along with several other features related to further integration with the VS Code API.

Reply:
I'm OK if we start with a highly 'opinionated' formatter (a-la Prettier) and only allow configuration away from the "one truth path" if users really scream about it. You keep the code (and the tests) way simpler the less config permutations you have.